### DISCUSSION TITLE
java.lang.ClassCastException: org.jgroups.stack.ProtocolStack cannot be cast to org.jgroups.protocols.TP

### DISCUSSION THREAD

**Discussion by Sagar-Singari** [belaban/JGroups#545] [Q&A]
Hi @belaban 

I keep getting errors while trying to upgrade from 3.3.3 to 3.6.10. Now, I'm getting this error:
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.informatica.b2b.MftServerBootstrap.main(MftServerBootstrap.java:92)
Caused by: **java.lang.ClassCastException: org.jgroups.stack.ProtocolStack cannot be cast to org.jgroups.protocols.TP**
	at org.jgroups.stack.ProtocolStack.getTransport(ProtocolStack.java:252)
	at org.jgroups.JChannel.stopStack(JChannel.java:1010)
	at org.jgroups.JChannel._close(JChannel.java:989)
	at org.jgroups.JChannel.close(JChannel.java:385)
	at com.linoma.dpa.jgroups.ClusterManager.close(ClusterManager.java:560)
	at com.linoma.startup.TomcatStarter.start(TomcatStarter.java:302)

Here's my Protocol stack:
                channel = new JChannel(false);
		channel.setName(SystemInfo.getName());
		channel.enableStats(false);
		ProtocolStack stack = new ProtocolStack();
		stack.enableStats(false);
		channel.setProtocolStack(stack);

                stack.addProtocol(getTCPProtocol());
                    if (initialMembers != null) {
                stack.addProtocol(getTCPPingProtocol());
                                                              }
		stack.addProtocol(getFDSockProtocol());
		stack.addProtocol(getFDAllProtocol());
		stack.addProtocol(getVerifySuspectProtocol());
		stack.addProtocol(getNakack2Protocol());
		stack.addProtocol(getUnicast3Protocol());
		stack.addProtocol(getCentralLockProtocol());
		stack.addProtocol(getStableProtocol());
		stack.addProtocol(getAuthProtocol());
		stack.addProtocol(getGMSProtocol());
		stack.addProtocol(getUfcProtocol());
		stack.addProtocol(getMfcProtocol());
		stack.addProtocol(getEncryptProtocol());
		stack.addProtocol(getFrag2Protocol());
        
		stack.init();
		channel.setDiscardOwnMessages(true);

		dispatcher = new MessageDispatcher(channel, null,
				new ClusterMembershipListener(), new ClusterMessageReceiver());

		lockService = new LockService(channel);

		channel.connect(ClusterConstants.CLUSTER_NAME);

Can you please help on this.
Thanks.

**Comment by belaban** [comment]
Do you have a reproducer that compiles under 3.6? Looks as if the transport is not of type TP...

**Reply by Sagar-Singari** [reply]
@belaban 
The code compiles fine with 3.6 version, I get this error only when I'm trying to cluster.
Do you think something could be wrong with the Protocol stack?

**Reply by belaban** [reply]
I don't know; that's why I need a reproducer

**Reply by Sagar-Singari** [reply]
@belaban 
The application is huge & the error occurs when I try to start my application in a cluster, but I'll still trying to refactor & get a reproducer.

Meanwhile, I had a question: Does the order of the protocols when defined in java using protocolStack.addProtocol() should be opposite to that as defined in XML? i.e in XML, if <TCP> is the 1st tag, then should it be added last when using  protocolStack.addProtocol(TCP) ?

**Reply by belaban** [reply]
No, the order is the same: the first protocol needs to be the transport. Take a look at TUNNEL_Test as example

**Reply by johnou** [reply]
@bobsag sounds like you have multiple versions of jgroups on your classpath.

**Reply by johnou** [reply]
You might try forcing a specific version in your project dependency management configuration.

**Reply by Sagar-Singari** [reply]
@johnou 
I already checked that, I'm using only a single version in classpath & also the same version is on the different instances on the cluster.

Thanks for your response!

**Reply by johnou** [reply]
Can you share the function close? com.linoma.dpa.jgroups.ClusterManager.close

**Reply by Sagar-Singari** [reply]
@johnou 
Here you go:
public void close() {
		ViewChangeRuntime.getInstance().shutdown();
		if (dispatcher != null || channel != null) {
			LOGGER.info(ClusterManagerMessages.getBundleString("info.clusterCloseStart"));
			if (channel != null) {
				channel.close();
				channel = null;
			}
			if (dispatcher != null) {
				dispatcher.stop();
				dispatcher = null;
			}
			LOGGER.info(ClusterManagerMessages.getBundleString("info.clusterCloseStop"));
		}
	}
